src/mame/barcrest/mpu1.cpp

```diff
@@ -0,0 +1,518 @@
+	machine().bookkeeping().coin_lockout_w(2, !state);
+	machine().bookkeeping().coin_lockout_w(3, !state);
+}
+
+TIMER_DEVICE_CALLBACK_MEMBER( mpu1_state::nmi )
+{
+	m_maincpu->pulse_input_line(INPUT_LINE_NMI, attotime::zero);
+}
+
+INPUT_CHANGED_MEMBER( mpu1_state::coin_input )
+{
+	switch(param)
+	{
+		case 0: m_pia1->ca1_w(newval); break;
+		case 1: m_pia1->cb1_w(newval); break;
+		case 2: m_pia2->ca1_w(newval); break;
+		case 3: m_pia2->cb1_w(newval); break;
+	}
+}
+
+TIMER_CALLBACK_MEMBER( mpu1_state::change_pia2a_bit7 )
+{
+	m_pia2a_bit7_value = 0;
+}
+
+/* MPU1 does not have stepper reels, it instead uses an electromechanical reel system.
+   Each reel has a single output - setting the output high causes the reel to start moving, 
+   and once it's set back low, the reel will stop at whichever symbol it's heading towards.
+   Position tracking is done via contacts on index plates below each symbol, which the CPU
+   will read once all reels are stopped.
+
+   I've modeled the reels as having 400 virtual "steps". Every reel has 20 symbols, which
+   gives 20 "steps" between each symbol. */
+void mpu1_state::reel_w(int reel, bool state)
+{
+	if(m_reel_state[reel] == REEL_STOPPED)
+	{
+		if(state)
+		{
+			m_reel_state[reel] = REEL_SPINNING;
+			m_reel_timer[reel]->adjust(m_reel_speed);
+			m_samples->play(fruit_samples_device::SAMPLE_EM_REEL_1_START + reel);
+		}
+	}
+	else if(m_reel_state[reel] == REEL_SPINNING)
+	{
+		if(!state)
+		{
+			if(m_reel_pos[reel] % 20 == 0) // If reel is already on a symbol, then stop it immediately
+			{
+				m_reel_timer[reel]->adjust(attotime::never);
+				m_reel_state[reel] = REEL_STOPPED;
+				m_samples->play(fruit_samples_device::SAMPLE_EM_REEL_1_STOP + reel);
+			}
+			else m_reel_state[reel] = REEL_STOPPING;
+		}
+	}
+}
+
+template <unsigned Reel> 
+TIMER_CALLBACK_MEMBER( mpu1_state::reel_move )
+{
+	if(m_reel_pos[Reel] == 0)
+		m_reel_pos[Reel] = 400 - 1;
+	else
+		m_reel_pos[Reel]--;
+
+	if(m_reel_state[Reel] == REEL_STOPPING && m_reel_pos[Reel] % 20 == 0) // Stop once a symbol is reached
+	{
+		m_reel_timer[Reel]->adjust(attotime::never);
+		m_reel_state[Reel] = REEL_STOPPED;
+		m_samples->play(fruit_samples_device::SAMPLE_EM_REEL_1_STOP + Reel);
+	}
+	else m_reel_timer[Reel]->adjust(m_reel_speed);
+
+	m_reel_out[Reel] = m_reel_pos[Reel];
+	m_sreel_out[Reel] = (m_reel_pos[Reel] * 0x10000) / 400;
+}
+
+static INPUT_PORTS_START( mpu1_inputs )
+	PORT_START("IN")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_UNUSED )
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_POKER_HOLD1 ) PORT_NAME("Hold 1")
+	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_POKER_HOLD2 ) PORT_NAME("Hold 2")
+	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_POKER_HOLD3 ) PORT_NAME("Hold 3")
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_POKER_CANCEL ) PORT_NAME("Cancel")
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_START1 ) PORT_NAME("Start")
+	PORT_BIT( 0x40, IP_ACTIVE_HIGH, IPT_UNUSED )
+	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_SERVICE ) PORT_NAME("Refill Key") PORT_CODE(KEYCODE_R) PORT_TOGGLE
```

Comment:
It’s better not to put custom names for inputs when the default names make sense.  It breaks localisation for one thing.  Also, it’s generally not a great idea to unnecessarily override the default mappings for inputs.  It makes it harder for people to set up controller configuration files, and it forces people to set per-game input assignments when they have a custom setup.

Reply:
Got rid of the custom names. The R for refill key thing seems to be standard across almost all the fruit drivers though, are overrides not favored at all? In that case making new input types seems like a good idea, usually there are multiple service inputs on fruit machines - refill key, doors, hopper dump switch, etc...

Reply:
Gambling inputs are still a bit of a mess, at some point I do intend to rework them a bit, but I don’t want to have to do it multiple times, so I want more time to think about it before I do it.  MAME does have multiple service input types that can be used for these things for now.